corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,323 @@
+            self.db,
+            self.original_doc_types,
+            self.iteration_key,
+            self.chunk_size,
+        )
+
+
+class ResumableDocsByTypeLogHandler(PaginateViewLogHandler):
+
+    def __init__(self, iterator):
+        self.iterator = iterator
+
+    def view_starting(self, db, view_name, kwargs, total_emitted):
+        offset = {k: v for k, v in kwargs.items() if k.startswith("startkey")}
+        self.iterator.state["offset"] = offset
+        self.iterator._save_state()
+
+
+class TooManyRetries(Exception):
+    pass
+
+
+DOCS_SKIPPED_WARNING = """
+        WARNING {} documents were not processed due to concurrent modification
+        during migration. Run the migration again until you do not see this
+        message.
+        """
+
+
+class BaseDocProcessor(six.with_metaclass(ABCMeta)):
+
+    def __init__(self, slug):
+        self.slug = slug
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    @abstractmethod
+    def process_doc(self, doc, couchdb):
+        """Migrate a single document
+
+        :param doc: The document dict to be migrated.
+        :param couchdb: Couchdb database in which to save migrated doc.
+        :returns: True if doc was migrated else False. If this returns False
+        the document migration will be retried later.
+        """
+        raise NotImplementedError
+
+    def processing_complete(self, skipped):
+        pass
+
+    def filter(self, doc):
+        """
+        :param doc: the document to filter
+        :return: True if this doc should be migrated
+        """
+        return True
+
+
+class CouchDocumentProcessor(object):
+    """Process Couch Docs
+
+    :param slug: process name.
+    :param doc_type_map: Dict of `doc_type_name: model_class` pairs.
+    :param doc_processor: A `BaseDocProcessor` object used to
+    process documents.
+    :param reset: Reset existing processor state (if any), causing all
+    documents to be reconsidered for processing, if this is true.
+    :param max_retry: Number of times to retry processing a document
+    before giving up.
+    :param chunk_size: Maximum number of records to read from couch at
+    one time. It may be necessary to use a smaller chunk size if the
+    records being processed are very large and the default chunk size of
+    100 would exceed available memory.
+    """
+    def __init__(self, doc_type_map, doc_processor, reset=False, max_retry=2, chunk_size=100):
+        self.doc_type_map = doc_type_map
+        self.doc_processor = doc_processor
+        self.reset = reset
+        self.max_retry = max_retry
+        self.chunk_size = chunk_size
+
+        self.couchdb = next(iter(doc_type_map.values())).get_db()
+        assert all(m.get_db() is self.couchdb for m in doc_type_map.values()), \
+            "documents must live in same couch db: %s" % repr(doc_type_map)
+
+        iter_key = doc_processor.slug + DOC_PROCESSOR_ITERATION_KEY_PREFIX
```

Comment:
Previously this was
```py
iter_key = slug + "-blob-migration"
```
Changing it will orphan existing migration states.

Also, it should be different for each type of iteration. It is possible that multiple different processes/migrations/etc. would iterate over the same set of doctypes. If two processes use the same slug and the same set of doc types then they will share the same iteration state. Now that this has been generalized that detail should be documented (probably makes sense to put it in the docstring).